File: client/src/types/srs.ts
// src/types/srs.ts
// SRS-specific types for the enhanced application

export type NodeStatus = 'fresh' | 'tackling' | 'grasped' | 'learned';
export type ReviewQuality = 0 | 1 | 2 | 3 | 4 | 5;
export type SessionType = 'definition' | 'exercise' | 'mixed';
export type SessionMode = 'normal' | 'frenzy';
export type SessionOrder = 'impact' | 'foundations';
export type DueView = 'full' | 'compact';
// Exercise solve state derived from solvedUntil/seenCount for UI coloring.
export type ExerciseSolveState = 'unsolved' | 'tried' | 'solved';

// Progress tracking for individual nodes
export interface NodeProgress {
  id: number;
  userId: number;
  nodeId: number;
  nodeType: 'definition' | 'exercise';
  nodeCode?: string;
  nodeName?: string;
  status: NodeStatus;
  easinessFactor: number;
  intervalDays: number;
  repetitions: number;
  lastReview?: string;
  nextReview?: string;
  accumulatedCredit: number;
  creditPostponed: boolean;
  totalReviews: number;
  successfulReviews: number;
  createdAt: string;
  updatedAt: string;
  // Derived fields used in UI — optional
  isDue?: boolean;
  daysUntilReview?: number | null;
  // Exercise-only solve tracking (solved while solvedUntil is in the future)
  solvedUntil?: string | null;
  seenCount?: number;
}

// Prerequisites relationship
export interface NodePrerequisite {
  id: number;
  nodeId: number;
  nodeType: 'definition' | 'exercise';
  prerequisiteId: number;
  prerequisiteType: 'definition' | 'exercise';
  weight: number; // 0.01 to 1.0
  createdAt: string;
}

// Study session tracking
export interface StudySession {
  id: number;
  userId: number;
  domainId: number;
  sessionType: SessionType;
  mode?: SessionMode;
  startTime: string;
  endTime?: string;
  totalReviews: number;
  successfulReviews: number;
  duration?: number; // in seconds
}

// Review submission. Success is derived server-side (quality >= 3).
export interface ReviewRequest {
  nodeId: number;
  nodeType: 'definition' | 'exercise';
  quality: ReviewQuality;
  timeTaken: number; // in seconds
  sessionId?: number;
  versionId?: number; // concrete version shown during the review
}

// Credit flow for animations
export interface CreditUpdate {
  nodeId: number;
  nodeType: 'definition' | 'exercise';
  credit: number;
  type: 'explicit' | 'implicit';
}

// Review response from API. counted=false means the node was not due and the
// grade was recorded as practice only (no SRS state changes).
export interface ReviewResponse {
  success: boolean;
  counted: boolean;
  message: string;
  updatedNodes?: NodeProgress[];
  creditFlow?: CreditUpdate[];
}

// Domain statistics
export interface DomainStats {
  domainId: number;
  totalNodes: number;
  freshNodes: number;
  tacklingNodes: number;
  graspedNodes: number;
  learnedNodes: number;
  dueReviews: number;
  completedToday: number;
  successRate: number;
}

export interface NotificationSummaryDomain {
  domainId: number;
  domainName: string;
  dueCount: number;
}

export interface NotificationSummaryInvite {
  id: number;
  domainId: number;
  domainName: string;
  invitedBy: number;
  invitedByUsername: string;
  role: 'editor' | 'viewer';
  createdAt: string;
}

export interface NotificationSummary {
  domains: NotificationSummaryDomain[];
  invites: NotificationSummaryInvite[];
  totalDue: number;
  inviteCount: number;
  generatedAt: string;
}

// Due review item
export interface DueReview {
  nodeId: number;
  nodeType: 'definition' | 'exercise';
  nodeCode: string;
  nodeName: string;
  status: NodeStatus;
  nextReview?: string;
  isDue: boolean;
  daysUntilReview?: number;
}

// Review queue item for practice/mixed sessions
export interface ReviewQueueItem {
  nodeId: number;
  nodeType: 'definition' | 'exercise';
  nodeCode: string;
  nodeName: string;
  isDue: boolean;
  exerciseMetaId?: number;
  exerciseMetaCode?: string;
  exerciseMetaName?: string;
}

// Review history item
export interface ReviewHistoryItem {
  id: number;
  userId: number;
  nodeId: number;
  nodeType: 'definition' | 'exercise';
  reviewTime: string;
  reviewType: 'explicit' | 'implicit';
  success: boolean;
  quality: ReviewQuality;
  timeTaken: number;
  creditApplied: number;
  easinessFactorBefore?: number;
  easinessFactorAfter?: number;
  intervalBefore?: number;
  intervalAfter?: number;
}

// UI state for animations
export interface CreditFlowAnimation {
  nodeId: string;
  credit: number;
  type: 'positive' | 'negative';
  timestamp: number;
}


// ==========================================================================
// Server-driven session engine
// ==========================================================================

export interface SessionStartRequest {
  domainId: number;
  sessionType: SessionType;
  mode?: SessionMode;
  order?: SessionOrder;
  exercisesPerDefinition?: number;
}

export interface SessionEngineItem {
  done: boolean;
  round: number;
  completed: number;
  correct: number;
  remaining: number;
  totalPlanned: number;
  item?: ReviewQueueItem;
  definitionVersion?: any;
  exerciseVersion?: any;
  lastReview?: ReviewResponse;
}

export interface SessionEngineState {
  session: StudySession;
  item: SessionEngineItem;
}

export interface SessionGradeRequest {
  quality: ReviewQuality;
  skip?: boolean;
  timeTaken?: number;
}

export function exerciseSolveState(progress: Pick<NodeProgress, 'solvedUntil' | 'seenCount'> | null | undefined): ExerciseSolveState {
  if (!progress) return 'unsolved';
  if (progress.solvedUntil && new Date(progress.solvedUntil).getTime() > Date.now()) {
    return 'solved';
  }
  if ((progress.seenCount ?? 0) > 0) {
    return 'tried';
  }
  return 'unsolved';
}

File: server/models/srs.go
package models

import (
	"time"
)

// NodePrerequisite represents a prerequisite relationship between nodes
type NodePrerequisite struct {
	ID               uint      `gorm:"primaryKey" json:"id"`
	NodeID           uint      `gorm:"column:node_id;not null" json:"nodeId"`
	NodeType         string    `gorm:"column:node_type;not null" json:"nodeType"` // 'definition' or 'exercise'
	PrerequisiteID   uint      `gorm:"column:prerequisite_id;not null" json:"prerequisiteId"`
	PrerequisiteType string    `gorm:"column:prerequisite_type;not null" json:"prerequisiteType"`
	Weight           float64   `gorm:"column:weight;default:1.0" json:"weight"`
	IsManual         bool      `gorm:"column:is_manual;default:false" json:"isManual"`
	CreatedAt        time.Time `gorm:"column:created_at;autoCreateTime" json:"createdAt"`
}

func (NodePrerequisite) TableName() string {
	return "node_prerequisites"
}

// UserNodeProgress represents SRS progress for a user on a specific node
type UserNodeProgress struct {
	ID                uint      `gorm:"primaryKey" json:"id"`
	UserID            uint      `gorm:"column:user_id;not null" json:"userId"`
	NodeID            uint      `gorm:"column:node_id;not null" json:"nodeId"`
	NodeType          string    `gorm:"column:node_type;not null" json:"nodeType"`
	Status            string    `gorm:"column:status;default:fresh" json:"status"` // fresh, tackling, grasped, learned
	EasinessFactor    float64   `gorm:"column:easiness_factor;default:2.5" json:"easinessFactor"`
	IntervalDays      float64   `gorm:"column:interval_days;default:0" json:"intervalDays"`
	Repetitions       int       `gorm:"column:repetitions;default:0" json:"repetitions"`
	LastReview        *time.Time `gorm:"column:last_review" json:"lastReview"`
	NextReview        *time.Time `gorm:"column:next_review" json:"nextReview"`
	AccumulatedCredit float64   `gorm:"column:accumulated_credit;default:0" json:"accumulatedCredit"`
	CreditPostponed   bool      `gorm:"column:credit_postponed;default:false" json:"creditPostponed"`
	TotalReviews      int       `gorm:"column:total_reviews;default:0" json:"totalReviews"`
	SuccessfulReviews int       `gorm:"column:successful_reviews;default:0" json:"successfulReviews"`
	CreatedAt         time.Time `gorm:"column:created_at;autoCreateTime" json:"createdAt"`
	UpdatedAt         time.Time `gorm:"column:updated_at;autoUpdateTime" json:"updatedAt"`
	
	// Relationships
	User *User `gorm:"foreignKey:UserID" json:"-"`
}

func (UserNodeProgress) TableName() string {
	return "user_node_progress"
}

// StudySession represents a study session
type StudySession struct {
	ID                uint      `gorm:"primaryKey" json:"id"`
	UserID            uint      `gorm:"column:user_id;not null" json:"userId"`
	DomainID          uint      `gorm:"column:domain_id;not null" json:"domainId"`
	SessionType       string    `gorm:"column:session_type;not null" json:"sessionType"` // definition, exercise, mixed
	StartTime         time.Time `gorm:"column:start_time;autoCreateTime" json:"startTime"`
	EndTime           *time.Time `gorm:"column:end_time" json:"endTime"`
	TotalReviews      int       `gorm:"column:total_reviews;default:0" json:"totalReviews"`
	SuccessfulReviews int       `gorm:"column:successful_reviews;default:0" json:"successfulReviews"`
	
	// Relationships
	User   *User   `gorm:"foreignKey:UserID" json:"-"`
	Domain *Domain `gorm:"foreignKey:DomainID" json:"-"`
}

func (StudySession) TableName() string {
	return "study_sessions"
}

// SessionReview represents an individual review within a session
type SessionReview struct {
	ID            uint      `gorm:"primaryKey" json:"id"`
	SessionID     uint      `gorm:"column:session_id;not null" json:"sessionId"`
	NodeID        uint      `gorm:"column:node_id;not null" json:"nodeId"`
	NodeType      string    `gorm:"column:node_type;not null" json:"nodeType"`
	ReviewType    string    `gorm:"column:review_type;not null" json:"reviewType"` // explicit, implicit
	ReviewTime    time.Time `gorm:"column:review_time;autoCreateTime" json:"reviewTime"`
	Success       bool      `gorm:"column:success;not null" json:"success"`
	Quality       *int      `gorm:"column:quality" json:"quality"` // 0-5, nullable for implicit reviews
	TimeTaken     *int      `gorm:"column:time_taken" json:"timeTaken"` // in seconds
	CreditApplied float64   `gorm:"column:credit_applied;default:1.0" json:"creditApplied"`
	
	// Relationships
	Session *StudySession `gorm:"foreignKey:SessionID" json:"-"`
}

func (SessionReview) TableName() string {
	return "session_reviews"
}

// ReviewHistory represents complete review history for analytics
type ReviewHistory struct {
	ID                    uint      `gorm:"primaryKey" json:"id"`
	UserID                uint      `gorm:"column:user_id;not null" json:"userId"`
	NodeID                uint      `gorm:"column:node_id;not null" json:"nodeId"`
	NodeType              string    `gorm:"column:node_type;not null" json:"nodeType"`
	ReviewTime            time.Time `gorm:"column:review_time;autoCreateTime" json:"reviewTime"`
	ReviewType            string    `gorm:"column:review_type;not null" json:"reviewType"`
	Success               bool      `gorm:"column:success;not null" json:"success"`
	Quality               *int      `gorm:"column:quality" json:"quality"`
	TimeTaken             *int      `gorm:"column:time_taken" json:"timeTaken"`
	CreditApplied         float64   `gorm:"column:credit_applied;default:1.0" json:"creditApplied"`
	EasinessFactorBefore  *float64  `gorm:"column:easiness_factor_before" json:"easinessFactorBefore"`
	EasinessFactorAfter   *float64  `gorm:"column:easiness_factor_after" json:"easinessFactorAfter"`
	IntervalBefore        *float64  `gorm:"column:interval_before" json:"intervalBefore"`
	IntervalAfter         *float64  `gorm:"column:interval_after" json:"intervalAfter"`
	
	// Relationships
	User *User `gorm:"foreignKey:UserID" json:"-"`
}

func (ReviewHistory) TableName() string {
	return "review_history"
}

// Review request/response models
type ReviewRequest struct {
	NodeID      uint   `json:"nodeId" binding:"required"`
	NodeType    string `json:"nodeType" binding:"required"`
	Success     bool   `json:"success"`
	Quality     int    `json:"quality" binding:"min=0,max=5"`
	TimeTaken   int    `json:"timeTaken"` // in seconds
	SessionID   *uint  `json:"sessionId"`
}

type ReviewResponse struct {
	Success        bool                    `json:"success"`
	Message        string                  `json:"message"`
	UpdatedNodes   []UserNodeProgress      `json:"updatedNodes,omitempty"`
	CreditFlow     []CreditUpdate         `json:"creditFlow,omitempty"`
}

type CreditUpdate struct {
	NodeID      uint    `json:"nodeId"`
	NodeType    string  `json:"nodeType"`
	Credit      float64 `json:"credit"`
	Type        string  `json:"type"` // explicit, implicit
}

// Session request/response models
type SessionRequest struct {
	DomainID    uint   `json:"domainId" binding:"required"`
	SessionType string `json:"sessionType" binding:"required"`
}

type SessionResponse struct {
	ID                uint      `json:"id"`
	DomainID          uint      `json:"domainId"`
	SessionType       string    `json:"sessionType"`
	StartTime         time.Time `json:"startTime"`
	EndTime           *time.Time `json:"endTime"`
	TotalReviews      int       `json:"totalReviews"`
	SuccessfulReviews int       `json:"successfulReviews"`
	Duration          *int      `json:"duration"` // in seconds
}

// Progress models
type NodeProgress struct {
	NodeID            uint       `json:"nodeId"`
	NodeType          string     `json:"nodeType"`
	NodeCode          string     `json:"nodeCode"`
	NodeName          string     `json:"nodeName"`
	Status            string     `json:"status"`
	EasinessFactor    float64    `json:"easinessFactor"`
	IntervalDays      float64    `json:"intervalDays"`
	Repetitions       int        `json:"repetitions"`
	LastReview        *time.Time `json:"lastReview"`
	NextReview        *time.Time `json:"nextReview"`
	AccumulatedCredit float64    `json:"accumulatedCredit"`
	CreditPostponed   bool       `json:"creditPostponed"`
	TotalReviews      int        `json:"totalReviews"`
	SuccessfulReviews int        `json:"successfulReviews"`
	DaysUntilReview   *int       `json:"daysUntilReview"`
	IsDue             bool       `json:"isDue"`
}

type DomainProgressSummary struct {
	DomainID           uint  `json:"domainId"`
	TotalNodes         int   `json:"totalNodes"`
	FreshNodes         int   `json:"freshNodes"`
	TacklingNodes      int   `json:"tacklingNodes"`
	GraspedNodes       int   `json:"graspedNodes"`
	LearnedNodes       int   `json:"learnedNodes"`
	DueReviews         int   `json:"dueReviews"`
	CompletedToday     int   `json:"completedToday"`
	SuccessRate        float64 `json:"successRate"`
}

// Status update request
type StatusUpdateRequest struct {
	NodeID   uint   `json:"nodeId" binding:"required"`
	NodeType string `json:"nodeType" binding:"required"`
	Status   string `json:"status" binding:"required"`
}

// Prerequisite models
type PrerequisiteRequest struct {
	NodeID           uint    `json:"nodeId" binding:"required"`
	NodeType         string  `json:"nodeType" binding:"required"`
	PrerequisiteID   uint    `json:"prerequisiteId" binding:"required"`
	PrerequisiteType string  `json:"prerequisiteType" binding:"required"`
	Weight           float64 `json:"weight"`
	IsManual         bool    `json:"isManual"`
}

type PrerequisiteResponse struct {
	ID               uint    `json:"id"`
	NodeID           uint    `json:"nodeId"`
	NodeType         string  `json:"nodeType"`
	PrerequisiteID   uint    `json:"prerequisiteId"`
	PrerequisiteType string  `json:"prerequisiteType"`
	PrerequisiteCode string  `json:"prerequisiteCode"`
	PrerequisiteName string  `json:"prerequisiteName"`
	Weight           float64 `json:"weight"`
	IsManual         bool    `json:"isManual"`
}
